src/plugins/hetero/src/plugin.cpp

```diff
@@ -77,23 +81,98 @@ ov::hetero::Plugin::DeviceProperties ov::hetero::Plugin::get_properties_per_devi
+}
+
 std::pair<ov::SupportedOpsMap, ov::hetero::SubgraphsMappingInfo> ov::hetero::Plugin::query_model_update(
     std::shared_ptr<ov::Model>& model,
     const ov::AnyMap& properties,
     bool allow_exception) const {
+    std::map<std::string, size_t> device_mem_map;
     Configuration full_config{properties, m_cfg};
     DeviceProperties properties_per_device =
         get_properties_per_device(full_config.device_priorities, full_config.get_device_properties());
 
     //  WARNING: Here is devices with user set priority
     auto device_names = ov::DeviceIDParser::get_hetero_devices(full_config.device_priorities);
+    bool hetero_query_model_by_device = false;
+    if (full_config.modelDistributionPolicy.count(ov::hint::ModelDistributionPolicy::PIPELINE_PARALLEL) != 0) {
+        hetero_query_model_by_device = get_device_memory_map(device_names, device_mem_map);
+    }
 
     auto update_supported_ops = [](ov::SupportedOpsMap& final_results, const ov::SupportedOpsMap& device_results) {
         for (const auto& layer_query_result : device_results)
             final_results.emplace(layer_query_result);
     };
 
+    auto has_subgraph_ops = [](std::shared_ptr<ov::Model>& model) {
+        for (auto& op : model->get_ordered_ops()) {
+            if (ov::as_type_ptr<ov::hetero::op::DeviceSubgraph>(op)) {
+                return true;
+            }
+        }
+        return false;
+    };
+
+    auto update_config = [&](ov::AnyMap& device_config,
+                             const std::shared_ptr<const ov::Model>& model,
+                             std::string device_name,
+                             bool fallback_device) {
+        auto supported_properties = get_core()->get_property(device_name, ov::supported_properties);
+        if (ov::util::contains(supported_properties, ov::query_model_ratio)) {
+            if (fallback_device) {
+                device_config[ov::query_model_ratio.name()] = 1.0f;
+            } else if (device_name.find("GPU") != std::string::npos &&
+                       device_mem_map.find(device_name) != device_mem_map.end()) {
+                size_t total_ops_size = 0;
+                for (auto&& op : model->get_ordered_ops()) {
+                    if (ov::op::util::is_constant(op)) {
+                        total_ops_size += op->get_element_type().size() * shape_size(op->get_shape());
+                    }
+                }
+                // Estimate the memory size required for the model is 1.2 * total_ops_size
+                // 1. Check if current GPU device that can take the entire model
+                // 2. Check if all left GPU devices can take the entire model
+                if (device_mem_map[device_name] >= 1.2 * total_ops_size) {
+                    device_config[ov::query_model_ratio.name()] = 1.0f;
+                } else if (device_mem_map["all_gpu_left"] >= 1.2 * total_ops_size ||
+                           device_mem_map.find("CPU") != device_mem_map.end()) {
```

Comment:
why do we have GPU hardcodes in common logic? can it be generalized on common discrete device?

Reply:
Updated, please review.